[fiber] Stackful cooperative scheduling via Fibers

salkinium opened the pull request:
This continues the work in #439. The changes/additions so far:

### Main function

The main function is not a fiber anymore, the developer has to manually start the scheduler. `modm::fiber::yield()` will return if scheduler is not running, therefore looping in the main functions. This will most likely work fine for interrupt/flags driven peripheral access like I2C, SPI and UART, however, won't work for any class that requires an update() function to be polled (but that didn't work with `RF_CALL_BLOCKING` either).

The `scheduler::run()` function will return to the main function if all fibers stop running (or if all functions have yielded). This can be used for sleeping and idle thread functionality.

This also treats the main function consistently across all platforms and carries no performance overhead if not using fibers.

I've also moved all classes into the `modm::fiber` namespace, since I'm worried about compatiblity with FreeRTOS and I don't want people believing that `modm::yield()` makes their code magically reentrant or something stupid like that. In general users should not have to manually yield, but rather use some higher-level primitive or modm API anyways.

### Cortex-M Context Switch

I've optimized the context switch assembly by using the push LR, pop PC pair correctly. This significantly speeds up the context switch. For FPU enabled devices, we only need to store the upper 16 floating point registers, since the lower ones are not saved across subfunction calls according to the EABI. This does double the switching time, perhaps in future we can investigate optimizations via the FPU flags register.

Fibers run on the PSP, while the main function and the interrupts use the MSP whose size is defined by the existing lbuild option `modm:platform:cortex-m:main_stack_size`.

### Guard Option

There is an invisible guard option `modm:__fibers` which prevents the fibers from being shown in modm.io modules and the doxygen docs until it is ready. This is particularly important for adding new peripheral drivers that may use a completely different API *without* resumable functions and shouldn't be exposed to the users right now.

### Stack Placement

On STM32F4 with CCM, the main stack has been moved into RAM, since the CCM is not DMA-able. On all other STM32 the main and `modm_faststack` are placed in the fastest DMA-able memory. We will have to see how well the Cortex-M7 DTCM is really accessible for DMA, we'll deal with that later.


TODO:

- ~Integrate [puncover](https://github.com/memfault/puncover) to get max stack size analysis.~ not trivial, will investigate later
- [x] AVR support
- [x] Cortex-M0 support
- [x] Cortex-M3 support
- [x] Cortex-M4/7 support
- [x] x86 support for Linux/macOS
- [x] x86 support for Windows
- ~ARM64 support~ Cannot test it, delegating to future
- [x] Guard option to hide fibers from users right now
- [x] Move all stacks (main + fibers) into the fastest DMA-able memory
- [x] Passing a lambda function and copying the lambda capture onto the stack
  - [x] Cortex-M
  - [x] AVR
  - [x] x86

rleh commented:
> The main stack will be reused for the IRQ stack, and the `scheduler::start()` function will return to the main function if all fibers stop running (or if all functions have yielded). This can be used for sleeping and idle thread functionality.

Shouldn't the function rather be named `scheduler::run()` (or `scheduler::update()`)?

salkinium commented:
Yes, but I'm still working to port fibers to AVR, and haven't gotten to the API yet.

salkinium commented:
I've added a context switch based on [avr-fibers](https://github.com/samukallio/avr-fiber) and the [GCC calling convention and register layout](https://gcc.gnu.org/wiki/avr-gcc#Register_Layout), however, it doesn't work entirely and as I cannot debug my AVR hardware, I've only been poking around in darkness as to why. In my mind it should just work, but I cannot even jump into the beginning of a fiber call. I'm going to use simavr to try and debug it.

salkinium commented:
I fixed the AVR fibers, turns out the ATmega2560 is using a 3 byte PC. 🙄
Validated the example and unit tests in hardware on ATmega2560 (3B PC) and ATmega328 (2B PC).

salkinium commented:
I'm not sure the x86_64 context switch is working correctly on Windows, I didn't understand what the previous implementation was doing with `%rax`. The examples and unittests work locally on macOS and in the Linux CI, but Windows seems to not like it? x86 seems to be a bit of a train wreck tho.

chris-durand commented:
I compared the x86_64 context switch to the implementation in boost context and found [big differences for the Microsoft ABI](https://github.com/boostorg/context/blob/develop/src/asm/jump_x86_64_ms_pe_gas.asm). They also save XMM6-XMM15 floating point registers and other things. It looks like we are also missing [some x87 floating point state](https://github.com/boostorg/context/blob/develop/src/asm/jump_x86_64_sysv_elf_gas.S#L37) for Unix x86_64 platforms. 

There is [some info](https://www.boost.org/doc/libs/1_77_0/libs/context/doc/html/context/rationale/x86_and_floating_point_env.html) in the boost context docs.

salkinium commented:
Oh boy… I'll check out if we can use setjmp/longjmp instead on Hosted. That would also give us ARM64 Linux support.

salkinium commented on examples/avr/fiber/main.cpp:
```diff
@@ -0,0 +1,90 @@
+modm::Fiber fiber1(stack1, &fiber_function1);
+modm::Fiber fiber2(stack2, &fiber_function2, (void*)cycles);
+modm::Fiber fiber3(stack3, &Test::fiber_function3, test);
+modm::Fiber fiber4(stack4, &Test::fiber_function4, test, (void*)cycles);
```
@chris-durand is there a better way to do this? Ideally without magical pointer/register foo?
I'm thinking about only accepting lambda functions and copying the lambda capture onto the stack and calling that lambda somehow (I think it's like what I do with the `this` pointer now). Wouldn't that work for all use-cases?

chris-durand replied:
A lambda with a capture is implementation-wise just syntactic sugar for an object of an anonymous struct type with an `operator()` defined and the captures are data members. Cppinsights can visualize [how the compiler sees a lambda](https://cppinsights.io/s/27e689ca). This is the the same as you are doing now for the member function case.

The implementation currently takes an l-value reference to an object and whoever creates the fiber has to guarantee it outlives the fiber. If you want to be able to pass an r-value
```
modm::Fiber some_fiber (stack, [foo, bar]() { … });
``` 
you have to store the lamba object somewhere.

chris-durand replied:
I really like the idea of using a lambda, `void*` are really ugly in a C++ api.

salkinium replied:
I'm stuck on the issue of getting a function pointer out of a lambda function. I cannot take the address of `operator()` since its not exposed. I've googled and found some really complicated solutions, but maybe you have a simpler solution.

salkinium commented on src/modm/processing/fiber/fiber.hpp:
```diff
@@ -0,0 +1,130 @@
+	::new ((void*)ptr) std::decay_t<T>{std::forward<T>(closure)};
+	// format the stack below the allocated closure
+	ctx = modm_context_init((uintptr_t) stack.memory, ptr,
+	                        (uintptr_t) &(decltype(closure)::operator()),
```
@chris-durand This line is the problem.

I'm copying the closure on the top of stack (% alignment) and then pass a pointer to this location to the lambda function as this pointer.

salkinium replied:
I solved it with a union. It's really ugly…

chris-durand replied:
There are some surprising corner cases with member function pointers. Generally you can't store a member function pointer in a `uintptr_t` and probably that's the reason the compiler doesn't let you do the cast. I guess it would compile if you cast to `void*` first.

For most platforms sizeof of a member function pointer is the size of a pointer plus a `ptrdiff_t`. If the class uses multiple inheritance the second half is the `this` pointer offset to the subclass. If the function is virtual, the first half is `0x1` and the second is the vtable offset to the member function.
More info:
 - https://devblogs.microsoft.com/oldnewthing/20040209-00/?p=40713
 - https://itanium-cxx-abi.github.io/cxx-abi/abi.html#member-function-pointers

Add that to the Linux fiber example and you will get a segfault:
```
struct Test3
{
        virtual void operator()()
        {
                MODM_LOG_INFO << "Test" << modm::endl;
                modm::fiber::yield();
        }
} test3;

modm::fiber::Stack<1024> stack3;
modm::Fiber fiber3(stack3, test3);
```
Do we care about this? It's a strange corner case that someone would use a virtual function in that case but it's ugly nevertheless.

The multiple inheritance subclass case works with gcc because it gives you the subclass member pointer type with a zero offset when you take the address to the function.

salkinium replied:
> Do we care about this?

No, the whole point of the lambda function is to avoid the specifics of the ABI particularly in complicated examples like this.
I'll check if I can enforce the use of a lambda function via Concepts.

chris-durand commented on src/modm/processing/fiber/fiber.hpp:
```diff
@@ -0,0 +1,151 @@
+	union {
+		decltype(&std::decay_t<T>::operator()) lambda = &std::decay_t<T>::operator();
+		uintptr_t function;
+	};
```
There is a pretty easy solution for handling the member function pointer ABI madness. Let the compiler generate the code for you:
```suggestion
        auto function = (uintptr_t) +[](std::decay_t<T>* closure) {
                (*closure)();
        };
```
This will work for the pointer to virtual function case.

salkinium replied:
Yup, that works great, simple and very nice, love it!

salkinium commented:
I cannot get the Windows context switch to work and I don't have the hardware available to debug this issue. I also don't fully understand why the context switch accesses the NT_TIB so I'm probably doing something very wrong. I've lost patience and don't care that much about Windows support for now.

chris-durand commented on src/modm/processing/fiber/context_x86_64.cpp:
```diff
@@ -0,0 +1,203 @@
+{
+	asm volatile
+	(
+		"popq %rdi		\n\t" // Load arg1
```
I would guess the Windows implementation fails here. Windows passes the first argument in RCX not RDI: https://en.wikipedia.org/wiki/X86_calling_conventions#Microsoft_x64_calling_convention

salkinium replied:
Right on the money!!!! It works now! \o/

chris-durand commented:
I'll review tomorrow.

salkinium commented:
I've moved all stacks into the fastest DMA-able memory and added a guard option `modm:__fibers` so that we can build a parallel, very much not backwards compatible API of peripheral drivers without confusing everyone. This also prevents the docs scripts from including the unfinished fibers docs.

We'll have to see how annoying it is to have two incompatible API next to each other…

I will also restructure the one commit into a few different ones, they are all squashed together until I got things working.

chris-durand commented on src/modm/processing/fiber/channel.hpp:
```diff
@@ -0,0 +1,105 @@
+	Data_t data;
+	Data_t* buffer = nullptr;
+	uint16_t buffer_size = 0;
+	uint16_t size;
```
`size` is uninitialized.
```suggestion
	uint16_t size = 0;
```

chris-durand commented on src/modm/processing/fiber/channel.hpp:
```diff
@@ -0,0 +1,105 @@
+	 *  @return The data from the channel.
+	 */
+	Data_t
+	recieve()
```
Spelling mistake
```suggestion
	receive()
```

chris-durand commented on src/modm/processing/fiber/context.h:
```diff
@@ -0,0 +1,44 @@
+
+#pragma once
+
+#include <stdint.h>
```
I'd use the C++ header instead of the C one.
```suggestion
#include <cstdint>
```

chris-durand commented on src/modm/processing/fiber/mutex.hpp:
```diff
@@ -0,0 +1,71 @@
+	void
+	acquire()
+	{
+		if (aquired) wait();
```
`aquired` should be spelled `acquired`

chris-durand commented on src/modm/processing/fiber/context_arm_m.cpp.in:
```diff
@@ -0,0 +1,204 @@
+ */
+// ----------------------------------------------------------------------------
+
+#include "context.h"
```
`context.h` doesn't compile as C, shouldn't it be renamed to `context.hpp` for consistency?

salkinium replied:
I've made it C compatible, but I'm not sure if it's useful to be able to call from a C context. We'll see.

chris-durand commented on src/modm/processing/fiber/waitable.hpp:
```diff
@@ -0,0 +1,89 @@
+	}
+
+private:
+	Fiber* lastWaiter;
```
This variable is uninitialized

chris-durand commented:
There is still something wrong with the `Waitable` implementation. Having two fibers block on an acquired mutex leads to weird behaviour and a segfault. I tested this code:
```
#include <modm/debug.hpp>
#include <modm/processing.hpp>

modm::fiber::Mutex m;

template<int index>
void test()
{
        for(int ii=0; ii<10; ii++)
        {
                m.acquire();
                MODM_LOG_INFO << "test" << index << "\n";
                modm::fiber::yield();
                m.release();
        }
}

modm::fiber::Stack<1024> stack[3];
modm::Fiber fiber1(stack[0], test<1>);
modm::Fiber fiber2(stack[1], test<2>);
modm::Fiber fiber3(stack[2], test<3>);

int
main(void)
{
        MODM_LOG_INFO << "Start" << modm::endl;
        modm::fiber::Scheduler::run();
        MODM_LOG_INFO << "End" << modm::endl;

        return 0;
}
```

It also crashes when I correctly initialize `Waitable::lastWaiter` as `nullptr`.

rleh reviewed (approved):
Nice!

salkinium commented:
I think the waitable implementation is too clever for me, it tries to reorder the list of fibers for low latency and I think it does it wrongly? 

I would prefer to have a dumber polling based implementation first to explore the usefulness of the API. The current implementation also doesn't work from an ISR context, which is what is needed to implement interrupt driven peripheral drivers.

I've removed the Waitables for now, since we also don't have a Protothread/Resumables version of those and therefore aren't necessary to replace them.

chris-durand reviewed (approved):
Let's merge it without the waitables. They can be added later if required.

salkinium committed:
[test] Split ATmega2560 tests into two parts

salkinium committed:
[hosted] Unify core peripheral type naming

salkinium committed:
[cortex-m] Remove unused process stack pointer

salkinium committed:
[cortex-m] Add .faststack section

salkinium committed:
[ci] Always execute all steps on macOS and Windows

salkinium committed:
[fiber] Add cooperative, stackful fibers

salkinium committed:
[examples] Add fiber examples

ghost commented:
Just an idea. If the fiber knows the stack size it only costs a few instructions extra to check for overflow during stack swap. It would be a nice option to have.
